src/libsystemd-network/dhcp-identifier.c

```diff
@@ -179,10 +185,19 @@ int dhcp_identifier_set_iaid(int ifindex, uint8_t *mac, size_t mac_len, void *_i
                 /* fall back to MAC address if no predictable name available */
                 id = siphash24(mac, mac_len, HASH_KEY.bytes);
 
-        id = htole64(id);
-
-        /* fold into 32 bits */
-        unaligned_write_be32(_id, (id & 0xffffffff) ^ (id >> 32));
+        id32 = (id & 0xffffffff) ^ (id >> 32);
+
+        if (legacy_unstable_byteorder) {
+                /* for historical reasons (a bug), the bits were swapped and thus
+                 * the result was endianness dependant. Preserve that behavior. */
+                id32 = __bswap_32(id32);
+        } else {
+                /* the fixed behavior returns a stable byte order. Since LE is expected
+                 * to be more common, swap the bytes on LE to give the same as legacy
+                 * behavior. */
+                id32 = be32toh(id32);
+        }
```

Comment:
i wonder if we really should retain compat here, and not instead sloppily just make the ordering used on LE cpus the default. I mean, BE is kinda dead anyway, no? hence we could just get away with not keeping compat here, and just doing it better for the future?

Reply:
(anyway, if this is important to you I'd also be willing to merge it as it is, but please, one nitpick: drop the {} around single-line if blocks, see CODING_STYLE)

Reply:
For NetworkManager's use, keeping the old behavior is not necessary because we soon won't call this function anymore. As a networkd user, I would find it annoying, if during package update the client-id changes. In particular, because downstream cannot really avoid it -- unless they are willing to patch their packages indefinitely to the old behavior (which seems too much effort).

Especially as the generated DUID depends on the machine-id, it is already host-specific, so in practice it is not important to be architecture independent. Probably nobody was badly affected by this, making a change annoying.

But I don't mind which solution is chosen. Either change behaviour for BE, or even don't do anything at all (because this patch currently doesn't actually change any behaviour).

---

I addressed the style issue and dropped the "for backward compatibility, generate iaid in native endianness." comments. The `legacy_unstable_byteorder` argument is sufficiently self-explanatory.